File: app/src/main/java/com/cashzhang/nozdormu/ObjectIO.java
package com.cashzhang.nozdormu;

import android.content.Context;
import android.util.Log;

import java.io.BufferedInputStream;
import java.io.BufferedOutputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.ObjectInput;
import java.io.ObjectInputStream;
import java.io.ObjectOutput;
import java.io.ObjectOutputStream;
import java.io.StreamCorruptedException;
import java.util.Collection;

/**
 * Created by cz21 on 2018/2/7.
 */


public class ObjectIO {
    private static final String TAG = ObjectIO.class.getSimpleName();
    private final Context m_context;
    private String m_fileName;
    private Object m_object;
    private int m_type;

    private static final int TYPE_FILE = 0;
    private static final int TYPE_COLLECTION = 1;
    private static final int TYPE_CONTENT = 2;

    public ObjectIO(Context context, String fileName, int type) {
        m_context = context;
        m_fileName = fileName;
        m_type = type;

    }
    public ObjectIO(Context context, int type) {
        m_context = context;
        m_type = type;
    }

    public void setNewFileName(String fileName) {
        m_fileName = fileName;
        m_object = null;
    }

    public boolean write(Object object) {
        try {
            StringBuilder stringBuilder = new StringBuilder();

            switch (m_type) {
                case TYPE_COLLECTION:
                    stringBuilder.append(m_context.getExternalFilesDir("collection"));
                    break;
                case TYPE_CONTENT:
                    stringBuilder.append(m_context.getExternalFilesDir("content"));
                    break;
                default:
                    break;
            }
            File file = new File(stringBuilder.toString()+"/"+m_fileName);
            Log.d(TAG, "write: filePath="+file.getAbsolutePath());
            if (!file.exists())
                file.createNewFile();
            ObjectOutputStream out = new ObjectOutputStream(new FileOutputStream(file));

            try {
                out.writeObject(object);
                return true;
            } finally {
                out.close();
            }
        } catch (FileNotFoundException e) {
            e.printStackTrace();
        } catch (IOException e) {
            e.printStackTrace();
        }
        return false;
    }

    int getElementCount() {
        int count = 0;
        try {
            if (null != m_object) {
                count = ((Collection) m_object).size();
            } else {
                Collection object = (Collection) read();
                if (null != object) {
                    count = object.size();
                }
            }
        } catch (ClassCastException ignored) {
        }
        return count;
    }

    public Object read() {
        try {
            StringBuilder stringBuilder = new StringBuilder();
            switch (m_type) {
                case 1:
                    stringBuilder.append(m_context.getExternalFilesDir("collection"));
                    break;
                case 2:
                    stringBuilder.append(m_context.getExternalFilesDir("content"));
                    break;
                default:
                    break;
            }
            File file = new File(stringBuilder.toString()+"/"+m_fileName);
            Log.d(TAG, "write: filePath="+file.getAbsolutePath());

            ObjectInput in = new ObjectInputStream(new FileInputStream(file));
            try {
                m_object = in.readObject();
                return m_object;
            } finally {
                in.close();
            }
        } catch (ClassNotFoundException e) {
        } catch (StreamCorruptedException e) {
        } catch (FileNotFoundException e) {
        } catch (IOException e) {
        }
        return null;
    }
}

File: app/src/main/java/com/cashzhang/nozdormu/ObjectIO.kt
package com.cashzhang.nozdormu

import android.content.Context
import android.util.Log
import com.cashzhang.nozdormu.ObjectIO
import java.io.*

/**
 * Created by cz21 on 2018/2/7.
 */
class ObjectIO {
    private val m_context: Context
    private var m_fileName: String? = null
    private var m_object: Any? = null
    private var m_type: Int

    constructor(context: Context, fileName: String?, type: Int) {
        m_context = context
        m_fileName = fileName
        m_type = type
    }

    constructor(context: Context, type: Int) {
        m_context = context
        m_type = type
    }

    fun setNewFileName(fileName: String?) {
        m_fileName = fileName
        m_object = null
    }

    fun write(`object`: Any?): Boolean {
        try {
            val stringBuilder = StringBuilder()
            when (m_type) {
                TYPE_COLLECTION -> stringBuilder.append(m_context.getExternalFilesDir("collection"))
                TYPE_CONTENT -> stringBuilder.append(m_context.getExternalFilesDir("content"))
                else -> {
                }
            }
            val file = File("$stringBuilder/$m_fileName")
            Log.d(TAG, "write: filePath=" + file.absolutePath)
            if (!file.exists()) file.createNewFile()
            val out = ObjectOutputStream(FileOutputStream(file))
            return try {
                out.writeObject(`object`)
                true
            } finally {
                out.close()
            }
        } catch (e: FileNotFoundException) {
            e.printStackTrace()
        } catch (e: IOException) {
            e.printStackTrace()
        }
        return false
    }

    val elementCount: Int
        get() {
            var count = 0
            try {
                if (null != m_object) {
                    count = (m_object as Collection<*>).size
                } else {
                    val `object` = read() as Collection<*>?
                    if (null != `object`) {
                        count = `object`.size
                    }
                }
            } catch (ignored: ClassCastException) {
            }
            return count
        }

    fun read(): Any? {
        try {
            val stringBuilder = StringBuilder()
            when (m_type) {
                1 -> stringBuilder.append(m_context.getExternalFilesDir("collection"))
                2 -> stringBuilder.append(m_context.getExternalFilesDir("content"))
                else -> {
                }
            }
            val file = File("$stringBuilder/$m_fileName")
            Log.d(TAG, "write: filePath=" + file.absolutePath)
            val `in`: ObjectInput = ObjectInputStream(FileInputStream(file))
            return try {
                m_object = `in`.readObject()
                m_object
            } finally {
                `in`.close()
            }
        } catch (e: ClassNotFoundException) {
        } catch (e: StreamCorruptedException) {
        } catch (e: FileNotFoundException) {
        } catch (e: IOException) {
        }
        return null
    }

    companion object {
        private val TAG = ObjectIO::class.java.simpleName
        private const val TYPE_FILE = 0
        private const val TYPE_COLLECTION = 1
        private const val TYPE_CONTENT = 2
    }
}